src/core/aio.c

```diff
@@ -556,16 +605,29 @@ nni_aio_expire_loop(void *arg)
 			continue;
 		}
 
+		// Find the timer with min expire time.
+		list = q->eq_list;
+		aio_idx = 0;
+		aio = list[aio_idx];
+		for (uint32_t i = 0; i < q->eq_len; i++) {
+			if (list[i]->a_expire < aio->a_expire) {
+				aio = list[i];
+				aio_idx = i;
+			}
+		}
 		if (now < aio->a_expire) {
-			// Unexpired; the list is ordered, so we just wait.
+			// Unexpired; we just wait the next expired aio.
 			nni_cv_until(cv, aio->a_expire);
 			now = nni_clock();
 			continue;
 		}
 
 		// The time has come for this aio.  Expire it, canceling any
 		// outstanding I/O.
-		nni_list_remove(list, aio);
+		for (uint32_t i = aio_idx+1; i < q->eq_len; i++) {
```

Comment:
this is silly.  the list is not ordered, so you don't need to do copy them all down.

you can just move the last element here instead.  this turns O(n) to just a single operation.

Reply:
btw, this can be done without a condition -- if this was the last item, then the swap will just move the item to itself, which won't matter, because you should then decrement the size.  (important: do the swap before you change the size.)  avoiding conditions is good because it means fewer pipeline stalls in the cpu.